api/rpc/stats/stats.go

```diff
@@ -74,12 +76,252 @@ func (s *Stats) doesElasticsearchServiceExist(ctx context.Context) bool {
+			return nil, err
+		}
+		if entry.Net.TxDropped, err = s.getInt64AvgValue(bucket, "avgTxDropped"); err != nil {
+			return nil, err
+		}
+		if entry.Net.TxErrors, err = s.getInt64AvgValue(bucket, "avgTxErrors"); err != nil {
+			return nil, err
+		}
+		if entry.Net.TxPackets, err = s.getInt64AvgValue(bucket, "avgTxPackets"); err != nil {
+			return nil, err
+		}
+	}
+	if req.StatsIo {
+		if entry.Io.Total, err = s.getInt64AvgValue(bucket, "avgIOTotal"); err != nil {
+			return nil, err
+		}
+		if entry.Io.Read, err = s.getInt64AvgValue(bucket, "avgIORead"); err != nil {
+			return nil, err
+		}
+		if entry.Io.Write, err = s.getInt64AvgValue(bucket, "avgIOWrite"); err != nil {
+			return nil, err
+		}
+	}
+	return entry, nil
+}
+
+func (s *Stats) statsCurrentQuery(ctx context.Context, req *StatsRequest) (*StatsReply, error) {
+	boolQuery := s.createBoolQuery(req, "now-10s")
+	agg := s.createTermAggreggation(req)
+
+	result, err := s.Client.Search().
+		Index(esIndex).
+		Query(boolQuery).
+		Aggregation("group", agg).
+		Size(0).
+		Do(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if result.Hits.TotalHits == 0 {
+		return &StatsReply{}, nil
+	}
+	ranges, ok := result.Aggregations.Terms("group")
+	if !ok {
+		return nil, fmt.Errorf("Request error 'group' not found")
+	}
+	ret := &StatsReply{}
+	for _, bucket := range ranges.Buckets {
+		entry, err := s.getEntryKey(req, bucket)
+		if err != nil {
+			return nil, err
+		}
+		ret.Entries = append(ret.Entries, entry)
+	}
+	return ret, nil
+}
+
+func (s *Stats) getFloat64AvgValue(bucket *elastic.AggregationBucketKeyItem, name string) (float64, error) {
+	avg, found := bucket.Avg(name)
+	if !found {
+		return 0, fmt.Errorf("Request error '%s' not found", name)
+	}
+	value := avg.Value
+	if value == nil {
+		return 0, nil
+	}
+	return *value, nil
+}
+
+func (s *Stats) getInt64AvgValue(bucket *elastic.AggregationBucketKeyItem, name string) (int64, error) {
+	val, err := s.getFloat64AvgValue(bucket, name)
+	if err != nil {
+		return 0, err
+	}
+	return int64(val), nil
+}
+
+func (s *Stats) statsHistoricQuery(ctx context.Context, req *StatsRequest) (*StatsReply, error) {
+	boolQuery := s.createBoolQuery(req, req.Period)
+	agg := s.createTermAggreggation(req)
+	histoAgg := elastic.NewDateHistogramAggregation().Field("@timestamp").Interval("30s")
+	agg = agg.SubAggregation("histo", histoAgg)
+
+	result, err := s.Client.Search().
+		Index(esIndex).
+		Query(boolQuery).
+		Aggregation("group", agg).
+		Size(0).
+		Do(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if result.Hits.TotalHits == 0 {
+		return &StatsReply{}, nil
+	}
+	ranges, ok := result.Aggregations.Terms("group")
+	if !ok {
+		return nil, fmt.Errorf("Request error 'group' not found")
+	}
+	ret := &StatsReply{}
+	for _, buck := range ranges.Buckets {
+		var found bool
+		histo, found := buck.Histogram("histo")
+		if !found {
+			panic("sub agg histo not found")
```

Comment:
Is panic something we want to be doing here? Not a microservice yet, so this currently will stop the whole `amplifier` process. I'm not quite clear, but is this even something that benefits from restarting the process?